kernel/kernel-api/src/main/java/io/delta/kernel/internal/commitrange/CommitRangeFactory.java

```diff
@@ -139,7 +147,137 @@ private void logResolvedVersions(long startVersion, Optional<Long> endVersionOpt
+   * version range, but does no validation that the list fills the range.
+   */
+  private List<ParsedDeltaData> getPublishedDeltasInVersionRange(
+      Engine engine, long startVersion, Optional<Long> endVersionOpt) {
+    final List<FileStatus> commitFiles =
+        listDeltaLogFilesAsIter(
+                engine,
+                Collections.singleton(FileNames.DeltaLogFileType.COMMIT),
+                tablePath,
+                startVersion,
+                endVersionOpt,
+                false /* mustBeRecreatable */)
+            .toInMemoryList();
+    List<ParsedDeltaData> publishedDeltas =
+        commitFiles.stream().map(ParsedDeltaData::forFileStatus).collect(Collectors.toList());
+
+    // Validate listed delta files are contiguous
+    if (publishedDeltas.size() > 1) {
+      for (int i = 1; i < publishedDeltas.size(); i++) {
+        final ParsedLogData prev = publishedDeltas.get(i - 1);
+        final ParsedLogData curr = publishedDeltas.get(i);
+        if (prev.getVersion() + 1 != curr.getVersion()) {
+          throw publishedDeltasNotContiguous(
+              tablePath.toString(),
+              publishedDeltas.stream()
+                  .map(ParsedDeltaData::getVersion)
+                  .collect(Collectors.toList()));
+        }
+      }
+    }
+    return publishedDeltas;
+  }
+
+  private void validatePublishedPlusRatifiedDeltas(
+      List<ParsedDeltaData> publishedDeltas, List<ParsedDeltaData> ratifiedDeltas) {
+    // Valid example: P0, P1, P2 + R1 (ratified within published)
+    // Valid example: P0, P1 + R2, R3 (no overlap)
+    // Valid example: P0, P1 + R1, R2 (overlap)
+    if (!publishedDeltas.isEmpty() && !ratifiedDeltas.isEmpty()) {
+      long earliestPublishedVersion = publishedDeltas.get(0).getVersion();
+      long earliestRatifiedVersion = ratifiedDeltas.get(0).getVersion();
+      // We cannot have ratifiedDeltas.head.version < publishedDeltas.head.version
+      // Example: P2, P3 + R1, R2, R3 is invalid
+      if (earliestRatifiedVersion < earliestPublishedVersion) {
+        throw catalogCommitsPrecedePublishedDeltas(
+            tablePath.toString(),
+            earliestRatifiedVersion,
+            publishedDeltas.stream().map(ParsedDeltaData::getVersion).collect(Collectors.toList()));
+      }
+      long lastPublishedVersion = ListUtils.getLast(publishedDeltas).getVersion();
+      // We must have publishedDeltas + ratifiedDeltas be contiguous
+      // Example: P0, P1 + R3, R4 is invalid
+      if (lastPublishedVersion + 1 < earliestRatifiedVersion) {
+        throw publishedDeltasAndCatalogCommitsNotContiguous(
+            tablePath.toString(),
+            publishedDeltas.stream().map(ParsedDeltaData::getVersion).collect(Collectors.toList()),
+            ratifiedDeltas.stream().map(ParsedDeltaData::getVersion).collect(Collectors.toList()));
+      }
+    }
+  }
+
+  private void validateVersionRange(
```

Comment:
nit: I think this method can be named `validateDeltasSatisfyRange` or sometihng like that? i.e. "to validate a version range" is a bit ambiguous